Add Publish All API

jdrossl opened the pull request:
https://github.com/craftercms/craftercms/issues/5591

jdrossl committed:
Add Publish All API

sumerjabri commented on src/main/java/org/craftercms/studio/impl/v2/repository/GitContentRepository.java:
```diff
@@ -1896,6 +1865,104 @@ private boolean createEnvironmentBranch(String siteId, String sandboxBranchName,
+                List<String> patterns = studioConfiguration.getList(CONFIGURATION_PUBLISHING_BLACKLIST_REGEX,
+                                                                             String.class);
+
+                List<String> ignored = status.getUncommittedChanges().stream()
```
Add a small comment to indicate you're deleting blacklisted files.

sumerjabri commented on src/main/java/org/craftercms/studio/impl/v2/repository/blob/s3/StudioAwsS3BlobStore.java:
```diff
@@ -361,4 +356,15 @@ public void initialPublish(String siteId) throws SiteNotFoundException {
+        Mapping targetMapping = getMapping(publishingTarget);
+
+        logger.debug("Performing publish all for site {0} to {1}", siteId, targetMapping);
+        deleteFolder(targetMapping.target, targetMapping.prefix, getClient());
```
This can be quite expensive. Can we not use AWS Sync mechanics?

jdrossl replied:
The Java SDK doesn't provide the `sync` feature from the CLI, the only option is to delete everything and copy it similar to what we do in the git repo

sumerjabri reviewed (changes requested):
I may have missed it, but where are we setting an audit record of publishAll?
Also, see notes in the files.

sumerjabri commented:
We agreed to pursue a manual sync using a change-set.
The publishAll will be split into a stage -> commit/rollback. Between staging and committing, we will sync the blob-store.

jdrossl committed:
Updates according to feedback:
- Split the publish all in two stages
- Avoid copying unchanged files for blob stores
- Record the operation in the audit log & activity stream

sumerjabri commented on src/main/java/org/craftercms/studio/impl/v2/repository/GitContentRepository.java:
```diff
@@ -1896,6 +1875,133 @@ private boolean createEnvironmentBranch(String siteId, String sandboxBranchName,
+            logger.info("Creating published repository for site {0}", siteId);
+            if (helper.createPublishedRepository(siteId, site.getSandboxBranch())) {
+                // return an empty set so no additional operations are performed
+                return new RepositoryChanges();
```
Where will we handle the blacklist?

sumerjabri replied:
Will be handled in the `createPublishedRepository` method in a future PR.

sumerjabri commented on src/main/java/org/craftercms/studio/impl/v2/repository/blob/BlobAwareContentRepository.java:
```diff
@@ -728,4 +729,44 @@ public void initialPublish(String siteId) throws SiteNotFoundException {
+            }
+            localRepositoryV2.completePublishAll(siteId, publishingTarget, gitChanges);
+        } catch (Exception e) {
+            logger.error("Error performing initial publish for site {0}", e, siteId);
```
Where do we unlock the git repo (generalized lock service)?

jdrossl committed:
Improvments for error handling during publish all

avasquez614 commented on src/main/java/org/craftercms/studio/impl/v2/repository/blob/s3/StudioAwsS3BlobStore.java:
```diff
@@ -361,4 +358,41 @@ public void initialPublish(String siteId) throws SiteNotFoundException {
+    }
+
+    @Override
+    public void completePublishAll(String siteId, String publishingTarget, RepositoryChanges changes) {
```
Wasn't there a way in S3 to bulk copy and delete?

jdrossl replied:
there is a bulk delete, but no bulk copy

avasquez614 replied:
do you think we would gain a performance improvement if we use the bulk delete?

jdrossl replied:
it might change a lot, probably there will be improvement only if a publish contains a lot of deleted files

avasquez614 replied:
@sumerjabri what do you think?

avasquez614 commented on src/main/java/org/craftercms/studio/impl/v2/service/publish/PublishServiceImpl.java:
```diff
@@ -230,6 +229,28 @@ public List<DeploymentHistoryGroup> getDeploymentHistory(String siteId, int days
+            throws ServiceLayerException, UserNotFoundException {
+        publishServiceInternal.publishAll(siteId, publishingTarget);
+
+        SiteFeed siteFeed = siteService.getSite(siteId);
```
Do we ignore audit records in case of an exception? If not this should be part of a finally

jdrossl replied:
as far as I know yeah, audit only records successful operations

jdrossl committed:
Use bulk delete for blob store during publish all
